### PR TITLE
sabnzbd: Add configurability of state directory owner and firewall

### LINKED ISSUE #276994: sabnzbd: Can only ever run as user called "sabnzbd"
Unfortunately the user and group configuration options do not carry through to the creation of the `/var/lib/sabnzbd` directory, as this is hardcoded to `sabnzbd` for the username. This directory has 700 permissions, meaning the service will not start if you try to run it as a different user.

https://github.com/NixOS/nixpkgs/blob/4ed0df40ea87766dda7ca5ff7095d9dd3cbd1370/nixos/modules/services/networking/sabnzbd.nix#L49C20-L49C27

### PR DESCRIPTION
**Pull request by cablespaghetti**
## Description of changes

Fixes #276994

Uses the systemd StateDirectory feature to manage `/var/lib/sabnzbd` and fixes the owner always being `sabnzbd` even if you have specified a different user/group to run the service.

Also adds an openFirewall option for good measure.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by ambroisie — CHANGES_REQUESTED** [review]
Please rebase and squash your commits (or even better: split them to atomic changes (`mkIf`-ying `user`/`group` creation, using `StateDirectory`, add `openFirewall`).

I'm not sure I see the usefulness of having `cfg.stateDirectoryMode`: who needs to look  at Sabnzbd's state files other than itself?

**Comment by cablespaghetti** [comment]
It took a couple of hours and my laptop nearly melted but I can now genuinely say that I've run the tests and they still pass. :laughing: 

I think this should be good to go now.

**Review by ambroisie — APPROVED** [review]

**Comment by cablespaghetti** [comment]
@adamcstephens I can see you've previously approved sabnzbd related PRs. Are you able to help get this merged? Thanks

**Comment by adamcstephens** [comment]
@ofborg test sabnzbd

File: nixos/modules/services/networking/sabnzbd.nix [hunk 1]
```diff
@@ -36,6 +36,14 @@ in
         default = "sabnzbd";
         description = lib.mdDoc "Group to run the service as";
       };
+
+      openFirewall = mkOption {
+        type = types.bool;
+        default = false;
+        description = lib.mdDoc ''
+          Open ports in the firewall for the sabnzbd web interface
+        '';
```
Comment (bjornfor) [inline nixos/modules/services/networking/sabnzbd.nix:45]: These descriptions should end with a dot.

**Pull request merged**